frame/merkle-mountain-range/src/tests.rs

```diff
@@ -347,28 +347,47 @@ fn should_verify() {
 }
 
 #[test]
-fn should_verify_batch_proof() {
+fn should_verify_batch_proofs() {
+	fn generate_and_verify_batch_proof(mmr_size: u64, leaves: &Vec<u64>) {
+		// (MMR Leafs)
+		let mut ext = new_test_ext();
+		ext.execute_with(|| add_blocks(mmr_size as usize));
+		ext.persist_offchain_overlay();
+
+		// Try to generate proof now. This requires the offchain extensions to be present
+		// to retrieve full leaf data.
+		register_offchain_ext(&mut ext);
+		let (leaves, proof) = ext.execute_with(|| {
+			crate::Pallet::<Test>::generate_batch_proof(leaves.to_vec()).unwrap()
+		});
+
+		ext.execute_with(|| {
+			add_blocks(7);
+			// then
+			assert_eq!(crate::Pallet::<Test>::verify_leaves(leaves, proof), Ok(()));
+		})
+	}
+
 	let _ = env_logger::try_init();
 
-	// Start off with chain initialisation and storing indexing data off-chain
-	// (MMR Leafs)
-	let mut ext = new_test_ext();
-	ext.execute_with(|| add_blocks(7));
-	ext.persist_offchain_overlay();
+	use itertools::Itertools;
+	for n in 1..8 {
+		// generate powerset (skipping empty set) of all possible leaf combinations for mmr size n
+		let leaves_set: Vec<Vec<u64>> = (0..n).into_iter().powerset().skip(1).collect();
 
-	// Try to generate proof now. This requires the offchain extensions to be present
-	// to retrieve full leaf data.
-	register_offchain_ext(&mut ext);
-	let (leaves, proof) = ext.execute_with(|| {
-		// when
-		crate::Pallet::<Test>::generate_batch_proof(vec![0, 4, 5]).unwrap()
-	});
+		leaves_set.iter().for_each(|leaves_subset| {
+			generate_and_verify_batch_proof(n, leaves_subset);
+		});
+	}
 
-	ext.execute_with(|| {
-		add_blocks(7);
-		// then
-		assert_eq!(crate::Pallet::<Test>::verify_leaves(leaves, proof), Ok(()));
-	});
+	for n in 8..15 {
+		// generate all possible 2-leaf combinations for mmr size n
+		let leaves_set: Vec<Vec<u64>> = (0..n).into_iter().combinations(2).collect();
```

Comment:
doesn't this end up repeating a good subset of combinations?

for `n=9` the test repeats all `(0..8)` combinations and so on and so forth 

disclaimer: I haven't checked what `combinations()` does _exactly_, so this might be stupid :see_no_evil: same question for `powerset()` loop

Reply:
You're correct: it does repeat them. It does so across different mmr sizes, i.e. with longer proofs, so it's testing distinct scenarios. Take for instance the case you reported in #11753 that verifying the leaves `[7, 11]` works for a 12 leaf mmr, but not a 13 leaf one.

`.combinations()` only does 2-leaf subsets here, whereas `.powerset()` does all subsets.